turbopuffer/namespace.py

```diff
@@ -199,178 +199,78 @@ def copy_from_namespace(self, source_namespace: str):
-               attributes: Optional[Dict[str, List[Optional[Union[str, int]]]]] = None,
-               schema: Optional[Dict] = None,
-               distance_metric: Optional[str] = None,
-               encryption: Optional[EncryptionDict] = None) -> None:
-        """
-        Creates or updates multiple vectors provided in a column-oriented layout.
-        If this call succeeds, data is guaranteed to be durably written to object storage.
+    def write(self,
+          upsert_columns: Optional[Dict[str, List]] = None,
+          patch_columns: Optional[Dict[str, List]] = None,
+          upsert_rows: Optional[List[Dict]] = None,
+          patch_rows: Optional[List[Dict]] = None,
+          deletes: Optional[List[Union[int, str]]] = None,
+          delete_by_filter: Optional[Filters] = None,
+          distance_metric: Optional[Literal["cosine_distance", "euclidean_squared"]] = None,
+          schema: Optional[Dict] = None,
+          encryption: Optional[EncryptionDict] = None) -> None:
+        """
+        Create, update, or delete documents.
+        """
+        payload = {}
+
+        if upsert_columns is not None:
+            if isinstance(upsert_columns, VectorColumns):
+                payload["upsert_columns"] = upsert_columns.to_dict_for_write()
+            elif isinstance(upsert_columns, Dict):
+                payload["upsert_columns"] = upsert_columns
+            else:
+                raise ValueError("upsert_columns must be a Dict or VectorColumns")
 
-        Upserting a vector will overwrite any existing vector with the same ID.
-        """
-        ...
+        if patch_columns is not None:
+            if isinstance(patch_columns, VectorColumns):
+                payload["patch_columns"] = patch_columns.to_dict_for_write()
+            elif isinstance(patch_columns, Dict):
+                payload["patch_columns"] = patch_columns
+            else:
+                raise ValueError("patch_columns must be a Dict or VectorColumns")
 
-    @overload
-    def upsert(self,
-               data: Union[dict, VectorColumns],
-               distance_metric: Optional[str] = None,
-               schema: Optional[Dict] = None,
-               encryption: Optional[EncryptionDict] = None) -> None:
-        """
-        Creates or updates multiple vectors provided in a column-oriented layout.
-        If this call succeeds, data is guaranteed to be durably written to object storage.
+        if upsert_rows is not None:
+            if not isinstance(upsert_rows, List):
+                raise ValueError("upsert_rows must be a List")
 
-        Upserting a vector will overwrite any existing vector with the same ID.
-        """
-        ...
+            if "upsert_columns" in payload:
+                raise ValueError("upsert_rows cannot be used with upsert_columns")
+            else:
+                payload["upsert_columns"] = VectorColumns.from_rows_for_write(upsert_rows).to_dict_for_write()
```

Comment:
this is a hack to convert row-based to column-based automatically, matching the python client's previous behaviour

not sure we really want/need that. and I just realized it's broken for patch

Reply:
it might also be best to get rid of these VectorRows and VectorColumns classes entirely. currently VectorRows is returned by the query and export functions but it adds some inflexibility and complexity

Reply:
by the way, this uncovered that the row-based APIs don't support base64 vectors at all. we'll want to fix that

Reply:
removed the automatic row->column transposition 

kept VectorRow and VectorColumns for now, but removed a lot of complex code